File: cli/src/commands/deploy.ts
import { Command } from 'commander'

import { getTokens, warnUnauthenticated } from '@common/auth'
import { Config } from '@common/types'
import { API_BASE_URL } from '@config'
import axios from 'axios'
import FormData from 'form-data'
import fs from 'fs'
import yaml from 'js-yaml'
import path from 'path'
import picomatch from 'picomatch'
import tar from 'tar'
import tmp from 'tmp'

const defaultIgnore = [
    '.git',
    '.gitmodules',
    '.vscode',
    '.next',
    '.npm',
    '.venv',
    '.yarn',
    '__pycache__',
    'bin',
    'dist',
    'node_modules',
    'yarn-error.log',
    '.DS_Store',
]

export const deploy = (program: Command) => {
    program
        .command('deploy')
        .description('Deploy a script')
        .action(async function () {
            const debug: boolean = this.opts().debug

            const tokens = await getTokens({ debug })
            const { access_token } = tokens

            // get current directory
            const cwd = process.cwd()
            debug && console.log(`Current directory: ${cwd}`)
            let folderToCheck = cwd

            let hasConfig = false

            // check each parent folder for a village.yaml file
            while (folderToCheck !== '') {
                debug &&
                    console.log(`Checking ${folderToCheck} for village.yaml`)
                const configPath = `${folderToCheck}/village.yaml`
                if (fs.existsSync(configPath)) {
                    debug &&
                        console.log(`Found village.yaml in ${folderToCheck}`)
                    hasConfig = true
                    break
                }
                const nextParent = folderToCheck
                    .split(path.sep)
                    .slice(0, -1) // remove last folder
                    .join(path.sep)
                if (nextParent === folderToCheck) {
                    break
                }
                folderToCheck = nextParent
            }

            if (!hasConfig) {
                console.error(
                    'No village.yaml file found. Run "village init" to create one'
                )
                return
            }

            const baseFolder = folderToCheck

            let ignoreContents: string[]

            try {
                ignoreContents = fs
                    .readFileSync(
                        `${baseFolder}{path.sep}.villageignore`,
                        'utf8'
                    )
                    .split('\n')
            } catch (err) {
                debug && console.log('No .villageignore file found')
                ignoreContents = []
            }

            // picomatch breaks on empty strings
            ignoreContents = ignoreContents.filter((line) => line.length > 0)

            // split ignored files into array
            const ignoredFiles = [...defaultIgnore, ...ignoreContents]

            const nonnegated = ignoredFiles.filter(
                (file) => !file.startsWith('!')
            )
            const negated = ignoredFiles
                .filter((file) => file.startsWith('!'))
                .map((file) => file.slice(1)) // remove the !

            const nonnegatedMatcher = picomatch(nonnegated)
            const negatedMatcher = picomatch(negated)

            const configFilePath = `${baseFolder}${path.sep}village.yaml`

            const configContents = fs.readFileSync(configFilePath, 'utf8')

            const config: Config = yaml.load(configContents) as Config

            const tmpFile = tmp.fileSync()
            debug &&
                console.log(
                    `Created temporary file ${tmpFile.name} for tarball`
                )

            await tar.c(
                {
                    gzip: true,
                    file: tmpFile.name,
                    filter: (name: string) => {
                        const keep =
                            !nonnegatedMatcher(name) || negatedMatcher(name)
                        debug &&
                            console.log(`${name} ${keep ? 'keep' : 'ignore'}`)
                        return keep
                    },
                },
                ['./']
            )

            const form = new FormData()
            form.append('script_id', config.id)
            form.append('context', fs.createReadStream(tmpFile.name))

            // use a raw axios request because openapi-generator doesn't like streams
            axios
                .post(`${API_BASE_URL}/script/build`, form, {
                    headers: {
                        ...form.getHeaders(),
                        Authorization: `Bearer ${access_token}`,
                    },
                })
                .then((res) => {
                    console.log('Script deployed successfully')
                    console.log(tmpFile.name)

                    // delete the temporary file
                    tmpFile.removeCallback()
                })
                .catch(warnUnauthenticated)
                .catch((err) => {
                    if (err.response.status === 404) {
                        const relativeConfigPath = path.relative(
                            cwd,
                            configFilePath
                        )
                        console.log(
                            `Script with id '${config.id}' does not exist!`
                        )
                        console.log(
                            `Create it by running 'village init --from ${relativeConfigPath}' first!`
                        )
                    }

                    if (err.response.status === 403) {
                        console.log(
                            `You do not have permission to deploy script with id '${config.id}'`
                        )
                    }

                    debug && console.log(err)

                    // delete the temporary file
                    tmpFile.removeCallback()
                })
        })
}

File: cli-rs/src/cli/deploy.rs
use crate::lib::auth::get_token;
use globset::{Glob, GlobSetBuilder};
use reqwest::header::AUTHORIZATION;
use reqwest::{multipart, Body, Client};
use serde::{Deserialize, Serialize};
use tempfile::NamedTempFile;
use tokio_util::codec::{BytesCodec, FramedRead};

use pathdiff;

use flate2::write::GzEncoder;
use flate2::Compression;

static DEFAULT_IGNORE: &'static [&'static str] = &[
    ".git",
    ".gitmodules",
    ".vscode",
    ".next",
    ".npm",
    ".venv",
    ".yarn",
    "__pycache__",
    "bin",
    "dist",
    "node_modules",
    "yarn-error.log",
    ".DS_Store",
];

#[derive(Debug, PartialEq, Serialize, Deserialize)]
struct Config {
    id: String,
}

pub async fn deploy() {
    let token = get_token();

    let cwd = std::env::current_dir().unwrap();

    // get config file in cwd or parents
    let mut folder_to_check = cwd.to_str().unwrap().to_owned();
    let mut has_config = false;

    while !has_config {
        let config_path = format!("{folder_to_check}/village.yaml");

        // check if village.yaml exists
        if std::path::Path::new(&config_path).exists() {
            has_config = true;
        } else {
            // check if we are at the root
            if folder_to_check == "/" {
                println!("No village.yaml found in this directory or any parent directory");
                return;
            }

            // go up a directory
            folder_to_check = std::path::Path::new(&folder_to_check)
                .parent()
                .unwrap()
                .to_str()
                .unwrap()
                .to_owned();
        }
    }

    let base_folder = folder_to_check;
    let config_path = format!("{base_folder}/village.yaml");
    let ignore_path = format!("{base_folder}/.villageignore");

    // read config
    let config_raw = std::fs::read_to_string(config_path).unwrap();
    let config: Config = serde_yaml::from_str(&config_raw).unwrap();

    let script_id = config.id;

    // read ignore file
    let ignore_contents = std::fs::read_to_string(ignore_path).unwrap();
    let ignore_contents = ignore_contents.split("\n").collect::<Vec<&str>>();
    let ignore = DEFAULT_IGNORE
        .iter()
        .chain(ignore_contents.iter())
        .map(|s| s.to_string())
        .collect::<Vec<String>>();

    // get negated and non-negated globs
    let nonnegated = ignore.iter().filter(|s| !s.starts_with("!"));
    let negated = ignore.iter().filter(|s| s.starts_with("!"));

    // build nonnegated matcher
    let mut nonnegated_matcher = GlobSetBuilder::new();
    for pattern in nonnegated {
        nonnegated_matcher.add(Glob::new(pattern).unwrap());
    }
    let nonnegated_matcher = nonnegated_matcher.build().unwrap();

    // build negated matcher
    let mut negated_matcher = GlobSetBuilder::new();
    for pattern in negated {
        negated_matcher.add(Glob::new(pattern).unwrap());
    }
    let negated_matcher = negated_matcher.build().unwrap();

    // compress files to temporary file
    let file = NamedTempFile::new().unwrap();
    let path = file.path();

    // build tarball
    let mut enc = GzEncoder::new(&file, Compression::default());
    {
        let mut tar = tar::Builder::new(&mut enc);

        // add all files to the tarball
        for entry in globwalk::GlobWalkerBuilder::from_patterns(&base_folder, &["**/*"])
            .build()
            .unwrap()
        {
            let entry = entry.unwrap();
            let path = entry.path().to_str().unwrap();
            let relative_path = pathdiff::diff_paths(path, &base_folder).unwrap();

            if !nonnegated_matcher.is_match(path) || negated_matcher.is_match(path) {
                tar.append_path_with_name(entry.path(), relative_path)
                    .unwrap();
            }
        }
    }

    // complete encoding
    enc.finish().unwrap();

    let file = tokio::fs::File::open(path).await.unwrap();

    // read file body stream
    let stream = FramedRead::new(file, BytesCodec::new());
    let file_body = Body::wrap_stream(stream);

    // create file stream
    let stream = multipart::Part::stream(file_body)
        .file_name("context.tar.gz")
        .mime_str("application/gzip")
        .unwrap();

    // creat form input
    let form = multipart::Form::new()
        .text("script_id", script_id)
        .part("context", stream);

    // post to api
    let res = Client::new()
        .post("http://localhost:8000/script/build")
        .header(AUTHORIZATION, format!("Bearer {}", token))
        .multipart(form)
        .send()
        .await;

    match res {
        Ok(r) => {
            println!("deployed");
        }
        Err(e) => {
            println!("error deploying: {:?}", e);
        }
    }
}
